src/ctx.cpp

```diff
@@ -43,7 +43,7 @@
 /*                             pj_get_ctx()                             */
 /************************************************************************/
 
-PJ_CONTEXT* pj_get_ctx( PJ *pj )
+PJ_CONTEXT* pj_get_ctx( const PJ *pj )
```

Comment:
This is fine for the compiler, but is it also logically right from an API design POV? 
`pj_get_ctx` may hand out a pointer to `pj->ctx`. And this pointer allows to modify the object pointed to by `pj`. In this sense, the function needs a pointer to a non-const object, even if the compiler is happy with a pointer to a const object.
IOW if you want a mutable `PJ_CONTEXT *`, you must bring a mutable `PJ *`.

Reply:
@dg0yt I'm not sure I share this analysis. ``pj_get_ctx( const PJ *pj )`` means that this particular function doesn't modify the passed PJ* object.  The fact that the returned PJ_CONTEXT* might be later modified, and those modifications could affect how the PJ* behaves is another concern.
If you look at the C standard library, there is a ``char *strchr(const char *s, int c);``that returns a char* pointer on the input const char*. But that's more because in C you can't have both ``char* strchr(char* s, int c);`` and ``const char* strchr(const char* s, int c);``
As pj_get_ctx() is in the internal C++ API, we could probably have 2 versions ``PJ_CONTEXT* pj_get_ctx(PJ* pj);`` and  ``const PJ_CONTEXT* pj_get_ctx(const PJ* pj);``

Reply:
`strchr` seems to really illustrate the point I'm making. Other than C, C++ offers two different overloads, for obvious reasons: https://en.cppreference.com/w/cpp/string/byte/strchr

Back in context. It is not about the inside (implementation) of the function, but about the outside (usage):
~~~c++
const PJ *my_const_pj = get_a_pj();
assert(has_some_property(my_const_pj));
do_something_fancy(pj_get_ctx(my_const_pj)); // Feeling safe, my_const_pj is const with no mutable properties.
assert(has_some_property(my_const_pj)); // Oops!
~~~

I don't have a strong opinion about it. I just would like to ensure that new compiler warnings don't translate too quickly into API changes.

Reply:
Whereas today, if you try to write a `do_something_fancy` that takes a const, you end up having to cast away the const when calling it, like: https://github.com/OSGeo/PROJ/pull/3496/commits/4dd98f60b73c2ef7c73de2762e413a113b013596

Reply:
@rouault These examples miss the point because they introduce direct access to the underlying object (`ctxt`, `v`) while I was explicitly illustrating a scope where there is only a pointer to the const object. FTR, I also didn't discuss concurrenct access.

@seanm The pointer to the actual usage patterns is useful. It seems like a common pattern is logging and setting errors, e.g.
~~~
proj_context_errno_set (pj_get_ctx ((PJ *) P), err);
~~~~~
Logging and error state change may be acceptable changes of observable state on an otherwise const object.

The calling scope has a `const PJ *`, and there is a compiler warning for `((PJ *) P)`. Let us enumerate all alternatives: 
- Make `pj_get_ctx` accept `const PJ *`.
  - Change the signature, and just benefit from the pointer to non-const (Your PR).
  - Add another overload.
- Permit this particular ignorance of `const` by an explicit C++ `const_cast<PJ *>(P) `.
  (In my example, it would make *visible* that something non-const happens between the two asserts.)
- Make `P` non-const.

Again, my point is not about const correctness on the compiler level. It ends at pointers. The compiler doesn't know our reasons for making member variables pointers instead of acutal types. My question is API correctness beyond such  implementation decisions.
And I don't know the answer. I do not know enough about the coupling between `PJ_CONTEXT` and `PJ`.

(I have seen big PRs guided by compiler warnings in other projects, and some of the initial changes were wrong.)

Reply:
>>As pj_get_ctx() is in the internal C++ API, we could probably have 2 versions PJ_CONTEXT* pj_get_ctx(PJ* pj); and const PJ_CONTEXT* pj_get_ctx(const PJ* pj);

>I could try that, see if it compiles...

It doesn't readily compile.

```C
int proj_errno_set (const PJ *P, int err) {
/******************************************************************************
    Set context-errno, bubble it up to the thread local errno, return err
******************************************************************************/
    /* Use proj_errno_reset to explicitly clear the error status */
    if (0==err)
        return 0;

    /* For P==0 err goes to the default context */
    proj_context_errno_set (pj_get_ctx (P), err); // <---------------- ••• error •••
    errno = err;

    return err;
}
```

The const-returning variant of `pj_get_ctx` gets passed to `proj_context_errno_set`, and it can't be changed to take const because it mutates the given `PJ_CONTEXT`.